src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -58,6 +59,93 @@ def check_type_refinement(node: foast.Expr, new: foast.DataType) -> None:
             )
 
 
+@dataclass
+class TypeInfo:
+
+    type: foast.SymbolType  # noqa: A003
+
+    @property
+    def is_complete(self) -> bool:
+        return is_complete_symbol_type(self.type)
+
+    @property
+    def is_any_type(self) -> bool:
+        if self.is_complete:
+            return False
+        return (self.type is None) or (self.type.constraint is None)
```

Comment:
Just a comment: have you thought about merging both conditions into a single `and` expression?

Reply:
It even looks to me like the first check is completely unnecessary.

Reply:
It isn't but I got everything on one line nonetheless